src/data/media-service.ts

```diff
@@ -0,0 +1,98 @@
+import { decodeMediaData } from '~/utils/decode-media-data'
+import { VersionedApiService } from '~/data/api-service'
+
+import type {
+  ApiQueryParams,
+  DetailFromMediaType,
+  FrontendMediaType,
+  MediaResult,
+  MediaStoreResult,
+} from '~/store/types'
+import type { SupportedMediaType } from '~/constants/media'
+
+import type { AxiosResponse } from 'axios'
+
+class MediaService<T extends FrontendMediaType> {
+  private readonly mediaType: T
+
+  constructor(mediaType: T) {
+    this.mediaType = mediaType
+  }
+
+  /**
+   * Decodes the text data to avoid encoding problems.
+   * Also, converts the results from an array of media objects into an object with
+   * media id as keys.
+   * @param data - search result data
+   */
+  transformResults(data: MediaResult<T[]>): MediaStoreResult<T> {
+    const mediaResults = <DetailFromMediaType<T>[]>data.results
+    return {
+      ...data,
+      results: mediaResults.reduce((acc, item) => {
+        acc[item.id] = decodeMediaData<T>(item, this.mediaType)
+        return acc
+      }, {} as T extends SupportedMediaType ? Record<string, DetailFromMediaType<T>> : never),
```

Comment:
This is incorrect, technically. It should just be `FrontendMediaType` as there are only two possible values for `T`, `image` and `audio` (unless there's also an ALL_CONTENT MediaService but it would fail to initialize as `this.mediaType` wouldn't match the `FrontendMediaType` constraint.

The only reason this is passing is because FrontendMediaType is a subset of SupportedMediaType, so it's saying, sure, it will extend SupportedMediaType.

Reply:
I have read and re-read your [first comment](https://github.com/WordPress/openverse-frontend/pull/1150#issuecomment-1073797171), and then just tried implementing it in the media conversion PR. That PR is quite large, but I suspect it's better to tackle these two things together because media depends on types in the services so much ...

Reply:
Could we split the PR to just typescriptify `api-service` maybe?